app/src/main/java/org/p2p/wallet/auth/ui/smsinput/NewSmsInputPresenter.kt

```diff
@@ -58,26 +68,53 @@ class NewSmsInputPresenter(
         }
 
         launch {
-            try {
-                view?.renderButtonLoading(isLoading = true)
-                createWalletInteractor.finishCreatingWallet(smsCode)
-                createWalletInteractor.finishAuthFlow()
-                view?.navigateToPinCreate()
-            } catch (incorrectSms: GatewayServiceError.IncorrectOtpCode) {
-                Timber.i(incorrectSms)
-                view?.renderIncorrectSms()
-            } catch (tooManyRequests: GatewayServiceError.TooManyRequests) {
-                Timber.i(tooManyRequests)
-                view?.navigateToSmsInputBlocked(GeneralErrorTimerScreenError.BLOCK_SMS_TOO_MANY_WRONG_ATTEMPTS)
-            } catch (serverError: GatewayServiceError.CriticalServiceFailure) {
-                Timber.e(serverError)
-                view?.navigateToCriticalErrorScreen(serverError.code)
-            } catch (error: Throwable) {
-                Timber.e(error, "Checking sms value failed")
-                view?.showUiKitSnackBar(messageResId = R.string.error_general_message)
-            } finally {
-                view?.renderButtonLoading(isLoading = false)
+            when (onboardingInteractor.currentFlow) {
+                OnboardingInteractor.OnboardingFlow.CREATE_WALLET -> finishCreatingWallet(smsCode)
+                OnboardingInteractor.OnboardingFlow.RESTORE_WALLET -> finishRestoringCustomShare(smsCode)
+            }
+        }
+    }
+
+    private suspend fun finishCreatingWallet(smsCode: String) {
+        try {
+            view?.renderButtonLoading(isLoading = true)
+            createWalletInteractor.finishCreatingWallet(smsCode)
+            createWalletInteractor.finishAuthFlow()
+            view?.navigateToPinCreate()
+        } catch (incorrectSms: GatewayServiceError.IncorrectOtpCode) {
+            Timber.i(incorrectSms)
+            view?.renderIncorrectSms()
+        } catch (tooManyRequests: GatewayServiceError.TooManyRequests) {
+            Timber.i(tooManyRequests)
+            view?.navigateToSmsInputBlocked(GeneralErrorTimerScreenError.BLOCK_SMS_TOO_MANY_WRONG_ATTEMPTS)
+        } catch (serverError: GatewayServiceError.CriticalServiceFailure) {
+            Timber.e(serverError)
+            view?.navigateToCriticalErrorScreen(serverError.code)
+        } catch (error: Throwable) {
+            Timber.e(error, "Checking sms value failed")
+            view?.showUiKitSnackBar(messageResId = R.string.error_general_message)
+        } finally {
+            view?.renderButtonLoading(isLoading = false)
+        }
+    }
+
+    private suspend fun finishRestoringCustomShare(smsCode: String) {
+        try {
+            view?.renderButtonLoading(isLoading = true)
+
+            restoreWalletRestoreInteractor.finishRestoreCustomShare(smsCode)
+
+            if (userRestoreInteractor.isUserReadyToBeRestored()) {
+                restoreUserWithShares()
```

Comment:
try to restore user using social share + custom share

request social share if it is absent